docs: how to manage remote systems (#547)
canonical/pebble: docs/how-to/manage-a-remote-system.md

# How to use Pebble to manage a remote system

> Managing server clusters remotely takes time and effort. Servers need regular updates, configuration changes, file transfers, and commands run. Directly managing remote servers or using tools like Ansible, which require SSH, increases overhead and security risks. The [XZ Utils backdoor incident](https://en.wikipedia.org/wiki/XZ_Utils_backdoor) highlighted vulnerabilities associated with SSH access.

Reviewer: This intro is great. A few minor suggestions for reading flow. Also I recommend mentioning the year of the XZ Utils backdoor:

```suggestion
Managing server clusters remotely takes time and effort. Servers need regular updates and configuration changes, which can involve transferring files and running commands. Directly managing remote servers or using tools that require SSH access (such as Ansible) increases overhead and security risks. The [XZ Utils backdoor incident](https://en.wikipedia.org/wiki/XZ_Utils_backdoor) in 2024 highlighted vulnerabilities associated with SSH access.
```

Pebble offers commands and an HTTP API over unix socket for remote system management, avoiding the need for extra open ports. 

> > Note: "remote system" means clients can interact with the Pebble daemon via a unix socket.

Reviewer: I suggest some extra detail. Not totally sure of this wording - what do you think?

```suggestion
> Note: By "remote system", we mean that the Pebble daemon is running in a separate system and there's a Unix socket for clients to interact with the daemon.
```

## Run commands in a remote system

One common task in system administration is updating and installing packages. With Pebble, we can use the `pebble exec` command to achieve this.

> For example, if Pebble is running as user with root privileges, we can use this command to update and install packages in a remote system:

Reviewer: "a user"?

```suggestion
For example, if Pebble is running as a user with root privileges, we can use this command to update and install packages in a remote system:
```

```{terminal}
:input: pebble exec apt update
Hit:1 http://ports.ubuntu.com/ubuntu-ports noble InRelease
Get:2 http://ports.ubuntu.com/ubuntu-ports noble-updates InRelease [126 kB]
Get:3 http://ports.ubuntu.com/ubuntu-ports noble-backports InRelease [126 kB]
...
Fetched 3121 kB in 38s (81.5 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
41 packages can be upgraded. Run 'apt list --upgradable' to see them.
```

To install a package, run:

```bash
pebble exec apt install cowsay
```

To confirm the package is successfully installed in the remote system, run:

```{terminal}
:input: pebble exec cowsay moo
 _____
< moo >
 -----
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

> For more information on the Pebble `exec` command, see {ref}`reference_pebble_exec_command`.

Reviewer: Very minor, but I'd probably shorten this to:
```suggestion
For more information, see {ref}`reference_pebble_exec_command`.
```

## Manage files in a remote system

Another common task in a remote system is configuration management. For example, updating configuration files, pushing files to servers, pulling files generated by running services, and so on.

With Pebble, this can be done easily. For example, if we want to install Nginx, create our own website, and serve it on a newly created virtual host, we can first install Nginx:

```bash
pebble exec apt install nginx
```

Then create our website _locally_:

```bash
echo "Hello, Pebble!" > /tmp/index.html
```

Create a directory in the remote system:

```bash
pebble mkdir -p /var/www/pebble
```

Push our local website file to the remote system:

```bash

> pebble push /tmp/index.html /var/www/pebble/index.html

Reviewer: Would it be possible to call the site `demo` (or similar) instead of `pebble`? That way I think it would be more obvious which parts are referring to the `pebble` command.

```

Create a virtual host configuration _locally_:

```
cat <<EOF > /tmp/pebble
server {
       listen 81;
       listen [::]:81;

       server_name example.ubuntu.com;

       root /var/www/pebble;
       index index.html;

       location / {
               try_files \$uri \$uri/ =404;
       }
}
EOF
```

Push the file to the remote system:

```bash
pebble push /tmp/pebble /etc/nginx/sites-enabled/pebble
```

Activate the newly added virtual host in the remote system by restarting Nginx:

```bash
pebble exec service nginx restart
```

> Finally, we can test the final result:

Reviewer: Suggest removing only having one "final" :slightly_smiling_face:

```suggestion
Finally, we can test the result:
```

```{terminal}
:input: curl localhost:81
Hello, Pebble!
```

For more information about related Pebble commands, see:

- {ref}`reference_pebble_ls_command`
- {ref}`reference_pebble_mkdir_command`
- {ref}`reference_pebble_rm_command`
- {ref}`reference_pebble_push_command`
- {ref}`reference_pebble_pull_command`

## Use the API

In the previous sections, we used Pebble commands to run commands and manage files. When automating these tasks, we can use the HTTP API that Pebble provides.

For example, to push a file to the remote system, we can POST to the `/v1/files` endpoint:

```bash
curl --unix-socket /path/to/.pebble.socket -XPOST http://_/v1/files -H "Content-Type: multipart/form-data" -F request='{"action": "write", "files": [{"path": "/var/www/pebble/index.html", "make-dirs": true, "permissions": "644"}]}' -F 'files=@/tmp/index.html;type=application/octet-stream;filename=/var/www/pebble/index.html'
```

> We can also use {ref}`api_go_client` and {ref}`api_python_client` for API access.

Reviewer: Suggest flipping these round:

```suggestion
We can also use the {ref}`api_python_client` and the {ref}`api_go_client` for API access.
```

## See more

> - [How to use the Pebble API to manage services](/how-to/use-the-pebble-api).
> - [API spec](/reference/api).

Reviewer: Suggestions for consistency:

```suggestion
- [How to use the Pebble API to manage services](/how-to/use-the-pebble-api)
- [API](/reference/api)
```

# How to use Pebble to manage a remote system

Managing server clusters remotely takes time and effort. Servers need regular updates and configuration changes, which typically involves transferring files and running commands. Directly managing remote servers or using tools that require SSH access (such as Ansible) increases overhead and security risks. The [XZ Utils backdoor incident](https://en.wikipedia.org/wiki/XZ_Utils_backdoor) in 2024 highlighted vulnerabilities associated with SSH access.

Pebble offers commands and an HTTP API over Unix socket for remote system management, avoiding the need for extra open ports. 

> Note: By "remote system", we mean that the Pebble daemon is running in a separate system and there's a Unix socket for clients to interact with the daemon.

## Run commands in a remote system

One common task in system administration is updating and installing packages. With Pebble, we can use the `pebble exec` command to achieve this.

> > Note: Set the environment variable `PEBBLE_SOCKET` to override the Unix socket used for the API (defaults to `$PEBBLE/.pebble.socket`).

Reviewer 2: ```suggestion
> Note: Set the environment variable `PEBBLE_SOCKET` to override the Unix socket used for the API (defaults to `$PEBBLE/.pebble.socket`, or to `/var/lib/pebble/default/.pebble.socket` if `PEBBLE` is not set).
```

With Pebble, this can be done easily. For example, if we want to install Nginx, create our own website, and serve it on a newly created virtual host, we can first install Nginx:

```bash
pebble exec apt install nginx
```

Then create our website _locally_:

```bash
echo "Hello, Pebble!" > /tmp/index.html
```

Create a directory in the remote system:

```bash
pebble mkdir -p /var/www/demo
```

Push our local website file to the remote system:

```bash
pebble push /tmp/index.html /var/www/demo/index.html
```

Create a virtual host configuration _locally_:

```
cat <<EOF > /tmp/demo
server {
       listen 81;
       listen [::]:81;

       server_name example.ubuntu.com;

       root /var/www/demo;
       index index.html;

       location / {
               try_files \$uri \$uri/ =404;
       }
}
EOF
```

Push the file to the remote system:

```bash
pebble push /tmp/demo /etc/nginx/sites-enabled/demo
```

Activate the newly added virtual host in the remote system by restarting Nginx:

```bash
pebble exec service nginx restart
```

Finally, we can test the result:

```{terminal}
:input: curl localhost:81
Hello, Pebble!
```

For more information about related Pebble commands, see:

> - {ref}`reference_pebble_ls_command`

Reviewer 2: How about we put these in alphabetical order? I can't see a particular order they're in now.

- {ref}`reference_pebble_mkdir_command`
- {ref}`reference_pebble_rm_command`
- {ref}`reference_pebble_push_command`
- {ref}`reference_pebble_pull_command`

## Use the API

In the previous sections, we used Pebble commands to run commands and manage files. When automating these tasks, we can use the HTTP API that Pebble provides.

For example, to push a file to the remote system, we can POST to the `/v1/files` endpoint:

```bash
curl --unix-socket /path/to/.pebble.socket -XPOST http://_/v1/files -H "Content-Type: multipart/form-data" -F request='{"action": "write", "files": [{"path": "/var/www/pebble/index.html", "make-dirs": true, "permissions": "644"}]}' -F 'files=@/tmp/index.html;type=application/octet-stream;filename=/var/www/pebble/index.html'
```

We can also use the {ref}`api_python_client` and the {ref}`api_go_client` for API access.

## See more

- [How to use the Pebble API to manage services](/how-to/use-the-pebble-api)

> - [API](/reference/api)

Reviewer 2: I'm not sure either of these links are particularly relevant. One is about managing services and the other just a generic link to the API reference. I think we could drop them.

How about we remove both of these sections and instead say something like, "You can also use the Pebble API to run command and manage files. This is normally done using the Go or Python client libraries. For more information, see [How to use the Pebble API](...)."
